src/weights.jl

```diff
@@ -2,44 +2,189 @@
 ###### Weight vector #####
 
 if VERSION < v"0.6.0-dev.2123"
-    immutable WeightVec{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
-        values::V
-        sum::S
-    end
+    @compat abstract type AbstractWeights{S<:Real, T<:Real, V<:RealVector} <: RealVector{T} end
 else
-    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
-        values::V
-        sum::S
+    @compat abstract type AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T} end
+end
+
+"""
+    @weights name
+
+generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
+and stores the `values` (`V<:RealVector`) and `sum` (`S<:Real`).
+"""
+macro weights(name)
+    return quote
+        if VERSION < v"0.6.0-dev.2123"
+            immutable $name{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        else
+            immutable $name{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
+                values::V
+                sum::S
+            end
+        end
     end
 end
 
+eltype(wv::AbstractWeights) = eltype(wv.values)
+length(wv::AbstractWeights) = length(wv.values)
+values(wv::AbstractWeights) = wv.values
+sum(wv::AbstractWeights) = wv.sum
+isempty(wv::AbstractWeights) = isempty(wv.values)
+
+Base.getindex(wv::AbstractWeights, i) = getindex(wv.values, i)
+Base.size(wv::AbstractWeights) = size(wv.values)
+
+"""
+    cfactor(n::Integer, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
+If `corrected=true` this will return ``\\frac{1}{n - 1}``
+(ie: [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise it will return ``\\frac{1}{n}``.
+"""
+cfactor(n::Integer, corrected=false) = 1 / (n - Int(corrected))
+
+"""
+    cfactor(wv::AbstractWeights, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with a set of
+weights `wv`.
+"""
+cfactor(wv::AbstractWeights, corrected=false) = cfactor(wv, Val{corrected})
+
+"""
+    cfactor(wv::AbstractWeights, false)
+
+``\\frac{1}{\sum w}``
+"""
+cfactor(wv::AbstractWeights, ::Type{Val{false}}) = 1 / sum(wv)
+cfactor(wv::AbstractWeights, ::Type{Val{true}}) =
```

Comment:
`Val` is really not what you want here: since the value of `corrected` will only be known at runtime, you're forcing the call to `cfactor` to go via dispatch at runtime, while it would have been inlined if you used `::Bool`.

Again, I don't think we should provide a fallback correction for `AbstractWeight`: it's unlikely to be valid, which is worse than raising an error. Creating new weight types shouldn't be a frequent need, people can afford the cost of defining this simple method if it applies.